pkg/urbit/jets/e/de_json.c

```diff
@@ -0,0 +1,1046 @@
+      byt_y = (c3_y*)vat_u->buf_w;
+    }
+    _init_stream(sam_u, byt_y, len_w);
+    _init_stack(tak_u);
+
+    //
+    // core logic
+    //
+
+    while (_json_peek(sam_u) != JSON_DONE) {
+      switch (_json_next(sam_u)) {
+        case JSON_ARRAY:
+        case JSON_OBJECT:
+        {
+          _push_stack(tak_u, col, tel);
+          col = u3_nul;
+          tel = u3_nul;
+          continue;
+        }
+
+        case JSON_ARRAY_END:
+        {
+          val = u3nc(c3__a, col);
+          _pop_stack(tak_u, &col, &tel);
+          break;
+        }
+
+        case JSON_OBJECT_END:
+        {
+          val = u3nc(c3__o, col);
+          _pop_stack(tak_u, &col, &tel);
+          break;
+        }
+
+        case JSON_STRING:
+        {
+          if ((_json_get_context(sam_u, &cnt_d) == JSON_OBJECT) && (cnt_d & 1)) {
+            key = u3i_string(_json_get_string(sam_u));
+            val = _DE_EMPTY_VAL;
+          } else {
+            val = u3nc(c3__s, u3i_string(_json_get_string(sam_u)));
+          }
+          break;
+        }
+
+        case JSON_NUMBER:
+        {
+          val = u3nc(c3__n, u3i_string(_json_get_string(sam_u)));
+          break;
+        }
+
+        case JSON_TRUE:
+        {
+          val = u3nc(c3__b, c3y);
+          break;
+        }
+
+        case JSON_FALSE:
+        {
+          val = u3nc(c3__b, c3n);
+          break;
+        }
+          
+        case JSON_NULL:
+        {
+          val = u3_nul;
+          break;
+        }
+
+        case JSON_ERROR:
+        {
+          _close(sam_u, tak_u);
+          return u3_nul;
+        }
+
+        default:
+        {
+          _close(sam_u, tak_u);
+          return u3m_bail(c3__fail);
+        }
+      }
+
+      switch (_json_get_context(sam_u, NULL)) {
+        case JSON_DONE:
+        {
+          res = val;
+          break;
+        }
+
+        case JSON_ARRAY:
+        {
+          u3_noun* nex;
+          u3_noun* hed;
+
+          if ( col == u3_nul ) {
+            nex = &col;
+          }
+          else {
+            nex = tel;
+          }
+
+          // TODO: does the root cell need to be incremented with u3k?
+          *nex = u3i_defcons(&hed, &tel);
+          *hed = val;
+          *tel = u3_nul;
+          break;
+        }
+
+        case JSON_OBJECT:
+        {
+          if (val != _DE_EMPTY_VAL) {
+            col = u3kdb_put(col, key, val);
+          }
+          break;
+        }
+      }
+    }
+
+    //
+    // clean up
+    //
+
+    c3_assert(res != u3_none && _json_peek(sam_u) == JSON_DONE);
+
+    _close(sam_u, tak_u);
+    return u3nc(u3_nul, res);
+  }
+
+  #undef _DE_EMPTY_VAL
+
+/* jet interface functions
+*/
+
+  u3_noun
+  u3qe_de_json(u3_atom a)
+  {
+    return _parse(a);
+  }
+
+  u3_noun
+  u3ke_de_json(u3_atom a)
+  {
+    u3_noun res = u3qe_de_json(a);
+    u3z(a);
+    return res;
+  }
+
+  u3_noun
+  u3we_de_json(u3_noun cor)
+  {
+    return u3qe_de_json(u3x_at(u3x_sam, cor));
+  }
```

Comment:
Can you try simply "vendoring" the upstream parser library by including it in this directory, and then forward-declaring the types/functions you need in this file?

Reply:
Done